### DISCUSSION TITLE
Users having trouble registering TOTP and "the token has already been consumed" log message

### DISCUSSION THREAD

**Discussion by aaww01** [authelia/authelia#6153] [Troubleshooting] [answered]
The following scenarios are playing out when users registering their TOTP device.
Looking for some suggestions or pointers on best way to address these issues:

**Scenario 1 - Register link confusion + gmail display**

1. User logs in with 1st factor, clicks register device link, waits too long to click the register link in the email, and clicks it after the token has expired, we see Token expired message in the log, and the user sees blurred QR code.  The register link appears to be valid for 5 minutes, and is hardcoded.
2. User logs in with 1st factor again, clicks register device link, but clicks the register link in the old email (Gmail, doesn't show the new email if it is identical to the previous email, it only shows the first email, of course, the register link in the wrong one.  we see Token expired message in the log, and the user sees blurred QR code.  The user gets stuck in this loop.

Suggested behavior: The email messages contains sufficiently distinct information from one message to the next to prevent gmail from collapsing them and only displaying the first instance.  For example, the "date/time" at which the message was generated and the "register link expires at hh:mm" might be a useful addition.   Users are in multiple timezones, so that might add some complexity. 

**Scenario 2 - the token has already been consumed.**

1. Builds on the first scenario where the user is repeatedly trying to (re)enroll the TOTP device
2. The user clicks on one of the previous registration emails but the token has not yet expired, or the user clicks a second time on the last message received after accidentally closing the browser tab, but received a blurred QR code and the log shows the token has already been consumed.  The end-user is confused.
3. Clicking the Register device link in the email immediately generates the event email Second Factor Method Added, despite the registration process not yet being completed.

Suggested behavior: Clicking the DONE button should be the signal to Authelia that the TOTP device has been registered and the token consumed, at this point the TOTP token is marked "active" internally in Authelia, and the email is sent.   Until then the TOTP token is just pending, and cannot be used to authenticate.

**Comment by james-d-elliott** [comment, marked as answer]
We're removing these links in favor of a single use randomly generated code. We do not have any plans to cater to any individual mail client however as it's likely an unwinnable battle when you take into consideration all of the possible clients.

Several of the other elements of this are already likely addressed in a WIP branch. See the following comment and associated discussion for updates on testing the functionality and providing feedback when the time comes: https://github.com/authelia/authelia/discussions/4813#discussioncomment-7284450

**Reply by aaww01** [reply]
I agree that anything that has to be done to deal with a specific email client is an unwinnable battle; the randomly generated code looks like it the right path to resolve that issue.  
Any plans to expand that to be one of the second factors?  I've come across a number of sites that have this option, and yes there are issues with that, email solutions that implement grey-listing is one.

Looking forward to the next release!

**Reply by james-d-elliott** [reply]
It's in our minds. It's been developed with this as a future optional expansion i.e. with reusable components and database tables. If we do it we're likely to require some sane security precautions by default i.e. that it's unavailable unless explicitly enabled, and that by default a user will be required to register a specific unique address for this purpose.

**Comment by james-d-elliott** [comment]
We added the code based option rather than links in a recent version 4.38 I believe.
